packages/core/src/amazonq/lsp/lspController.ts

```diff
@@ -0,0 +1,378 @@
+import fetch from 'node-fetch'
+import { LspClient } from './lspClient'
+import AdmZip from 'adm-zip'
+import { RelevantTextDocument } from '@amzn/codewhisperer-streaming'
+import { makeTemporaryToolkitFolder } from '../../shared/filesystemUtilities'
+import { CodeWhispererSettings } from '../../codewhisperer/util/codewhispererSettings'
+import { activate as activateLsp } from './lspClient'
+import { telemetry } from '../../shared/telemetry'
+import { isCloud9 } from '../../shared/extensionUtilities'
+import { globals } from '../../shared'
+import { AuthUtil } from '../../codewhisperer'
+import { isWeb } from '../../shared/extensionGlobals'
+function getProjectPaths() {
+    const workspaceFolders = vscode.workspace.workspaceFolders
+    if (!workspaceFolders || workspaceFolders.length === 0) {
+        throw Error('No workspace folders found')
+    }
+    return workspaceFolders.map(folder => folder.uri.fsPath)
+}
+
+export interface Chunk {
+    readonly filePath: string
+    readonly content: string
+    readonly context?: string
+    readonly relativePath?: string
+    readonly programmingLanguage?: string
+}
+
+export interface Content {
+    filename: string
+    url: string
+    hashes: string[]
+    bytes: number
+    serverVersion?: string
+}
+
+export interface Target {
+    platform: string
+    arch: string
+    contents: Content[]
+}
+
+export interface Manifest {
+    manifestSchemaVersion: string
+    artifactId: string
+    artifactDescription: string
+    isManifestDeprecated: boolean
+    versions: {
+        serverVersion: string
+        isDelisted: boolean
+        targets: Target[]
+    }[]
+}
+const manifestUrl = 'https://aws-toolkit-language-servers.amazonaws.com/temp/manifest.json'
+// this LSP client in Q extension is only going to work with these LSP server versions
+const supportedLspServerVersions = ['0.1.0']
+
+const nodeBinName = process.platform === 'win32' ? 'node.exe' : 'node'
+/*
+ * LSP Controller manages the status of Amazon Q LSP:
+ * 1. Downloading, verifying and installing LSP using DEXP LSP manifest and CDN.
+ * 2. Managing the LSP states. There are a couple of possible LSP states:
+ *    Not installed. Installed. Running. Indexing. Indexing Done.
+ * LSP Controller converts the input and output of LSP APIs.
+ * The IDE extension code should invoke LSP API via this controller.
+ * 3. It perform pre-process and post process of LSP APIs
+ *    Pre-process the input to Index Files API
+ *    Post-process the output from Query API
+ */
+export class LspController {
+    static #instance: LspController
+    private _isIndexingInProgress = false
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+    constructor() {}
+
+    isIndexingInProgress() {
+        return this._isIndexingInProgress
+    }
+
+    async _download(localFile: string, remoteUrl: string) {
+        const res = await fetch(remoteUrl, {
```

Comment:
can the shared `packages/core/src/common/request.ts` be used instead?

Reply:
No it does not pipe to a file writestream. 

I got resp.body?.pipeTo is not a function

Reply:
But I am using request.fetch when getting the manifest.